### PR TITLE
JAMES-2586 Implement PostgresEmailChangeRepository

### PR DESCRIPTION
**Pull request by hungphan227**

### CODE REVIEW

File: server/data/data-jmap-postgres/src/main/java/org/apache/james/jmap/postgres/change/PostgresStateFactory.java [hunk 1]
```diff
@@ -0,0 +1,31 @@
...
+package org.apache.james.jmap.postgres.change;
+
+import org.apache.james.jmap.api.change.State;
+
+import com.github.f4b6a3.uuid.UuidCreator;
+
+public class PostgresStateFactory implements State.Factory {
+    @Override
+    public State generate() {
+        return State.of(UuidCreator.getTimeBased());
```
Comment (chibenwa) [inline server/data/data-jmap-postgres/src/main/java/org/apache/james/jmap/postgres/change/PostgresStateFactory.java:29]: Let's use uuid v7 ?

See https://datatracker.ietf.org/doc/draft-ietf-uuidrev-rfc4122bis/

```
   2.  The 100-nanosecond Gregorian epoch used in UUIDv1 (described in
       Section 5.1) timestamps is uncommon and difficult to represent
       accurately using a standard number format such as [IEEE754].

   3.  Introspection/parsing is required to order by time sequence, as
       opposed to being able to perform a simple byte-by-byte
       comparison.
       
   4.  Privacy and network security issues arise from using a MAC
       address in the node field of UUID version 1.  Exposed MAC
       addresses can be used as an attack surface to locate network
       interfaces and reveal various other information about such
       machines (minimally manufacturer, potentially other details).
       Additionally, with the advent of virtual machines and containers,
       MAC address uniqueness is no longer guaranteed.
```

File: server/data/data-jmap-postgres/src/main/java/org/apache/james/jmap/postgres/change/PostgresEmailChangeModule.java [hunk 2]
```diff
@@ -0,0 +1,64 @@
...
+import org.jooq.impl.SQLDataType;
+
+public interface PostgresEmailChangeModule {
+    interface PostgresEmailChangeTable {
+        Table<Record> TABLE_NAME = DSL.table("email_change");
+
+        Field<String> ACCOUNT_ID = DSL.field("account_id", SQLDataType.VARCHAR.notNull());
+        Field<UUID> STATE = DSL.field("state", SQLDataType.UUID.notNull());
+        Field<OffsetDateTime> DATE = DSL.field("date", SQLDataType.TIMESTAMPWITHTIMEZONE.notNull());
+        Field<Boolean> IS_DELEGATED = DSL.field("is_delegated", SQLDataType.BOOLEAN.notNull());
```
Comment (vttranlina) [inline server/data/data-jmap-postgres/src/main/java/org/apache/james/jmap/postgres/change/PostgresEmailChangeModule.java:42]: We can change the name to `is_shared`

tldr: In the grooming ticket, I copied DDL table from cassandra and forgot it.
In pass, we had a ticket for renaming it 
https://github.com/apache/james-project/pull/1605
https://github.com/linagora/tmail-backend/pull/729#discussion_r1228922984

File: server/data/data-jmap-postgres/src/main/java/org/apache/james/jmap/postgres/change/PostgresEmailChangeModule.java [hunk 3]
```diff
@@ -0,0 +1,64 @@
...
+        Field<UUID[]> CREATED = DSL.field("created", SQLDataType.UUID.getArrayDataType().notNull());
```
Comment (vttranlina) [inline server/data/data-jmap-postgres/src/main/java/org/apache/james/jmap/postgres/change/PostgresEmailChangeModule.java:43]: Just curious, it should be always not null?
I checked the ChangeBuilder, it was initialed by List.empty, 
But it just a code, not a real case logic

**Review by Arsnael — COMMENTED** [review]
Read it, nothing to add

**Review by quantranhong1999 — COMMENTED** [review]
Nothing to add. Just a reminder as we are missing TTL handling.

File: server/data/data-jmap-postgres/src/main/java/org/apache/james/jmap/postgres/change/PostgresEmailChangeModule.java (hunk 3 above)
Comment (hungphan227) [reply server/data/data-jmap-postgres/src/main/java/org/apache/james/jmap/postgres/change/PostgresEmailChangeModule.java:43]: Because the List created by ChangeBuilder is not null, I set these fields not null :))

**Comment by hungphan227** [comment]
> Nothing to add. Just a reminder as we are missing TTL handling.

Created another ticket for this: https://github.com/linagora/james-project/issues/5040

**Comment by Arsnael** [comment]
Don't we need an index though on that timestamp column to manage efficiently the TTLs even with a external cronjob on the side?

**Comment by hungphan227** [comment]
> Don't we need an index though on that timestamp column to manage efficiently the TTLs even with a external cronjob on the side?

added

**Review by Arsnael — APPROVED** [review]

**Review by quantranhong1999 — APPROVED** [review]

**Comment by hungphan227** [comment]
missing bindings

**Comment by hungphan227** [comment]
> missing bindings

added

**Review by Arsnael — APPROVED** [review]

**Review by vttranlina — APPROVED** [review]
